# Студенти у Републици Србији

У овој радној свесци ћемо отворене податке o студентима прерадити и припремити мању агрегирану табелу за даљу обраду и визуализације. Сирове податке преузели смо са сајта Министарства за просвету, науку и технолошки развој: http://opendata.mpn.gov.rs/index.php?ucenici_studenti=visoko. Фајл у .csv формату се налази у фолдеру raw data. У наставку ћемо користећи библиотеке *pandas* и *cyrtranslit*: 
- трансформисати табелу (додавати и уклањати неке редове и колоне, груписати податке, селектовати само универзитетске студенте и сл.)
- преписати податке из ћириличног у латинично писмо
- сачуваћемо процесирану табелу за даљу обраду у следећим табелама.
Ова радна свеска је замишљена као додатни материјал за оне који се интересују за то како је оригинална табела са сајта МПНТР-а преточена у мању табелу коришћену у остатку анализе. Предлажемо да прво прођете кроз радну свеску са обрадом и визуализацијом ових података, а тек онда да се упустите у овај мало досаднији посао прераде података.

In [1]:
import pandas as pd
import cyrtranslit

Као и претходно користићемо библиотеку **pandas** за учитавање и преглед података, док ћемо у овој свесци користити и библиотеку **cyrtranslit** за превођење ћириличног у латинично писмо.
Подаци су преузети са поменуте адресе и сачувани у локалном фолдеру, ако преузмете неку новију верзију података, промените наредну команду:

In [2]:
studenti = pd.read_csv('raw data/MPNTRopendata_vs.csv')
studenti.head(2)

#  ИД установе               Универзитет  \
0  1.0           82    Универзитет у Приштини   
1  2.0           62  Универзитет у Новом Саду   

                        Назив установе      Тип установе  ИД програма  \
0  Правни факултет, Косовска Митровица  Државни факултет           16   
1             Факултет техничких наука  Државни факултет          616   

               Назив програма                                      Звање  \
0                         ОАС                        Дипломирани правник   
1  Инжењерски менаџмент (SSS)  Специјалиста струковни инжењер менаџмента   

                           Поље               Ниво  ...  \
0  Друштвено-хуманистичке науке  Основне академске  ...   
1     Техничко-технолошке науке    Спец. струковне  ...   

   Број самофинансирајући студената - шеста година  \
0                                              NaN   
1                                              NaN   

   Број страних студената - прва година  \
0                                   0.0   
1                                   0.0   

   Број страних студената - друга година  \
0                                    0.0   
1                                    NaN   

  Број страних студената - трећа година  \
0                                   0.0   
1                                   NaN   

  Број страних студената - четврта година  \
0                                     0.0   
1                                     NaN   

   Број страних студената - пета година  \
0                                   NaN   
1                                   NaN   

   Број страних студената - шеста година  \
0                                    NaN   
1                                    NaN   

   Број свршених студената током претходне школске године  \
0                                                 42        
1                                                 22        

   Број пријављених кандидата на упису у наредну школску годину - I рок  \
0                                              104.0                      
1                                               29.0                      

   Број пријављених кандидата на упису у наредну школску годину - II рок  
0                                                0.0                      
1                                                0.0                      

[2 rows x 37 columns]

Ова табела о студентима садржи различите високошколске програме, за сваки имамо податке о универзитету, нивоу студија, бројевима студената, звању које студенти стичу и слично. Како је овај преглед колона скраћен, остатак доступних колона можемо видети уз помоћ функције **columns**:

In [3]:
studenti.columns

Index(['#', 'ИД установе', 'Универзитет', 'Назив установе', 'Тип установе',
       'ИД програма', 'Назив програма', 'Звање', 'Поље', 'Ниво',
       'Школарина за домаће студенте', 'Школарина за стране студенте',
       'Број ЕСПБ бодова', 'Број решења акредитације',
       'Датум решења акредитације', 'Акредитациона квота',
       'Број буџетских студената - прва година',
       'Број буџетских студената - друга година',
       'Број буџетских студената - трећа година',
       'Број буџетских студената - четврта година',
       'Број буџетских студената - пета година',
       'Број буџетских студената - шеста година',
       'Број самофинансирајући студената - прва година',
       'Број самофинансирајући студената - друга година',
       'Број самофинансирајући студената - трећа година',
       'Број самофинансирајући студената - четврта година',
       'Број самофинансирајући студената - пета година',
       'Број самофинансирајући студената - шеста година',
       'Број страних студе

Пре даље анализе преписаћемо имена колона у латинично писмо ради лакшег куцања кода и позивања колона у наставку. То радимо уз помоћ функције **to_latin** из библиотеке *cyrtranslit*:

In [4]:
studenti.columns = [cyrtranslit.to_latin(text,'sr') for text in list(studenti.columns)]

Користећи функцију **info** излистаћемо све колоне да извидимо тип података који се у њима крије као и број уноса који постоји у њима:

In [5]:
studenti.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1924 entries, 0 to 1923
Data columns (total 37 columns):
#                                                                         1924 non-null float64
ID ustanove                                                               1924 non-null int64
Univerzitet                                                               1501 non-null object
Naziv ustanove                                                            1924 non-null object
Tip ustanove                                                              1924 non-null object
ID programa                                                               1924 non-null int64
Naziv programa                                                            1924 non-null object
Zvanje                                                                    1924 non-null object
Polje                                                                     1924 non-null object
Nivo                                   

Видимо у табели да су нам на располагању бројеви студената на буџету, самофинансирању и страни студенти по годинама, стога ћемо увести колоне у којима су и укупни бројеви студената све ове три категорије. У сабирању ових колона користићемо функцију **sum**:

In [6]:
studenti['Broj budžetskih studenata'] = studenti[['Broj budžetskih studenata - prva godina', 'Broj budžetskih studenata - druga godina','Broj budžetskih studenata - treća godina','Broj budžetskih studenata - četvrta godina','Broj budžetskih studenata - peta godina','Broj budžetskih studenata - šesta godina']].sum(axis=1)
studenti['Broj samofinansirajućih studenata'] = studenti[['Broj samofinansirajući studenata - prva godina', 'Broj samofinansirajući studenata - druga godina','Broj samofinansirajući studenata - treća godina','Broj samofinansirajući studenata - četvrta godina','Broj samofinansirajući studenata - peta godina','Broj samofinansirajući studenata - šesta godina']].sum(axis=1)
studenti['Broj stranih studenata'] = studenti[['Broj stranih studenata - prva godina','Broj stranih studenata - druga godina','Broj stranih studenata - treća godina','Broj stranih studenata - četvrta godina','Broj stranih studenata - peta godina','Broj stranih studenata - šesta godina']].sum(axis=1)
studenti['Broj studenata'] = studenti[['Broj budžetskih studenata','Broj samofinansirajućih studenata','Broj stranih studenata']].sum(axis=1)

За сваки од студијских програма видимо да табела садржи и информацију о броју ЕСПБ бодова, које можемо превести у дужину трајања програма по семестрима и годинама (користећи чињеницу да школска година има 60, а семестар 30 ЕСПБ бодова):

In [7]:
studenti['Dužina trajanja programa (u semestrima)'] = studenti['Broj ESPB bodova']//30
studenti['Dužina trajanja programa (u god)'] = studenti['Broj ESPB bodova']//60

Потражња за одређеним факултетом дата је у две колоне, кроз два уписна рока, стога ћемо додати и колону у којој ћемо чувати укупан број пријављених кандидата:

In [8]:
studenti['Broj prijavljenih kandidata'] = studenti[['Broj prijavljenih kandidata na upisu u narednu školsku godinu - I rok','Broj prijavljenih kandidata na upisu u narednu školsku godinu - II rok']].sum(axis=1)

Зарад испитивања пролазности студената кроз различите академске године, направићемо и колоне у којима су студенти сваке од година, независно да ли су у питању студенти на буџету или не:

In [9]:
studenti['Studenti prva godina'] = studenti[['Broj budžetskih studenata - prva godina','Broj samofinansirajući studenata - prva godina','Broj stranih studenata - prva godina']].sum(axis=1)
studenti['Studenti druga godina'] = studenti[['Broj budžetskih studenata - druga godina','Broj samofinansirajući studenata - druga godina','Broj stranih studenata - druga godina']].sum(axis=1)
studenti['Studenti treća godina'] = studenti[['Broj budžetskih studenata - treća godina','Broj samofinansirajući studenata - treća godina','Broj stranih studenata - treća godina']].sum(axis=1)
studenti['Studenti četvrta godina'] = studenti[['Broj budžetskih studenata - četvrta godina','Broj samofinansirajući studenata - četvrta godina','Broj stranih studenata - četvrta godina']].sum(axis=1)
studenti['Studenti peta godina'] = studenti[['Broj budžetskih studenata - peta godina','Broj samofinansirajući studenata - peta godina','Broj stranih studenata - peta godina']].sum(axis=1)
studenti['Studenti šesta godina'] = studenti[['Broj budžetskih studenata - šesta godina','Broj samofinansirajući studenata - šesta godina','Broj stranih studenata - šesta godina']].sum(axis=1)

Високо образовање груписано је у пар поља која можемо видети излиставањем јединствених вредности у овој колони користећи функцију **unique**. *На исти начин можете истражити и остале текстуалне колоне које вас занимају, испробајте!*

In [10]:
studenti.Polje.unique()

array(['Друштвено-хуманистичке науке', 'Техничко-технолошке науке',
       'Природно-математичке науке',
       'Интердисциплинарне, мултидисциплинарне, трансдисциплинарне (ИМТ) и двопредметне студије',
       'Уметност', 'Медицинске науке'], dtype=object)

Често ћемо желети да цртамо графике који могу бити обојени или означени образовним пољима којима припадају студијски програми, стога нам је згодно да скратимо њихове називе, то можемо урадити следећом функциојом:

In [11]:
def mapiranjepolja(text):
    temp=''
    if text=='Друштвено-хуманистичке науке':
        temp='DH'
    if text=='Техничко-технолошке науке':
        temp='TT'
    if text=='Природно-математичке науке':
        temp='PM'
    if text=='Интердисциплинарне, мултидисциплинарне, трансдисциплинарне (ИМТ) и двопредметне студије':
        temp='IMT'
    if text=='Уметност':
        temp='UM'
    if text=='Медицинске науке':
        temp='MD'
    return(temp)

Користећи функцију **apply** лако можемо применити једноставну функцију, као ову коју смо претходно направили (*mapiranjepolja*) на сваки елемент неке колоне:

In [12]:
studenti['Polje_skraceno'] = studenti.Polje.apply(mapiranjepolja)
studenti.head(2)

#  ID ustanove               Univerzitet  \
0  1.0           82    Универзитет у Приштини   
1  2.0           62  Универзитет у Новом Саду   

                        Naziv ustanove      Tip ustanove  ID programa  \
0  Правни факултет, Косовска Митровица  Државни факултет           16   
1             Факултет техничких наука  Државни факултет          616   

               Naziv programa                                     Zvanje  \
0                         ОАС                        Дипломирани правник   
1  Инжењерски менаџмент (SSS)  Специјалиста струковни инжењер менаџмента   

                          Polje               Nivo  ...  \
0  Друштвено-хуманистичке науке  Основне академске  ...   
1     Техничко-технолошке науке    Спец. струковне  ...   

   Dužina trajanja programa (u semestrima)  Dužina trajanja programa (u god)  \
0                                        8                                 4   
1                                        2                                 1   

   Broj prijavljenih kandidata Studenti prva godina Studenti druga godina  \
0                        104.0                129.0                 182.0   
1                         29.0                 40.0                   0.0   

   Studenti treća godina  Studenti četvrta godina  Studenti peta godina  \
0                  211.0                    432.0                   0.0   
1                    0.0                      0.0                   0.0   

   Studenti šesta godina  Polje_skraceno  
0                    0.0              DH  
1                    0.0              TT  

[2 rows x 51 columns]

Слично можемо видети и које универзитете имамо међу подацима:

In [13]:
studenti['Univerzitet'].unique()

array(['Универзитет у Приштини', 'Универзитет у Новом Саду',
       'Универзитет у Нишу', 'Универзитет у Београду',
       'Универзитет у Крагујевцу', 'Универзитет уметности у Београду',
       'Државни Универзитет у Новом Пазару', nan,
       'Универзитет Сингидунум', 'Универзитет Унион - Никола Тесла',
       'Универзитет Мегатренд', 'Универзитет одбране',
       'Универзитет Едуконс', 'Универзитет Алфа БК'], dtype=object)

Како имамо у подацима и неке од приватних универзитета згодно је увести колону власништва која може згодно доћи у наставку анализе:

In [14]:
def mapiranjeuniverziteta(text):
    if text in ['Универзитет у Приштини', 'Универзитет у Новом Саду',
       'Универзитет у Нишу', 'Универзитет у Београду',
       'Универзитет у Крагујевцу', 'Универзитет уметности у Београду',
       'Државни Универзитет у Новом Пазару', 'Универзитет одбране']:
        return 'Државни факултет'
    elif text in ['Универзитет Сингидунум', 'Универзитет Унион - Никола Тесла',
       'Универзитет Мегатренд', 'Универзитет Едуконс', 'Универзитет Алфа БК']:
        return 'Приватни факултет'
    else:
        return ''

In [15]:
studenti['Vlasnistvo']=studenti['Univerzitet'].apply(mapiranjeuniverziteta)
studenti.head(2)

#  ID ustanove               Univerzitet  \
0  1.0           82    Универзитет у Приштини   
1  2.0           62  Универзитет у Новом Саду   

                        Naziv ustanove      Tip ustanove  ID programa  \
0  Правни факултет, Косовска Митровица  Државни факултет           16   
1             Факултет техничких наука  Државни факултет          616   

               Naziv programa                                     Zvanje  \
0                         ОАС                        Дипломирани правник   
1  Инжењерски менаџмент (SSS)  Специјалиста струковни инжењер менаџмента   

                          Polje               Nivo  ...  \
0  Друштвено-хуманистичке науке  Основне академске  ...   
1     Техничко-технолошке науке    Спец. струковне  ...   

   Dužina trajanja programa (u god)  Broj prijavljenih kandidata  \
0                                 4                        104.0   
1                                 1                         29.0   

   Studenti prva godina Studenti druga godina Studenti treća godina  \
0                 129.0                 182.0                 211.0   
1                  40.0                   0.0                   0.0   

   Studenti četvrta godina  Studenti peta godina  Studenti šesta godina  \
0                    432.0                   0.0                    0.0   
1                      0.0                   0.0                    0.0   

   Polje_skraceno        Vlasnistvo  
0              DH  Државни факултет  
1              TT  Државни факултет  

[2 rows x 52 columns]

In [16]:
studenti['Vlasnistvo'].unique()

array(['Државни факултет', '', 'Приватни факултет'], dtype=object)

Додатним увидом можемо проверити шта су програми код којих смо власништво оставили празно, и испоставља се да су у питању високе школе:

In [17]:
studenti[studenti.Vlasnistvo=='']

#  ID ustanove Univerzitet  \
253    254.0          127         NaN   
254    255.0          127         NaN   
255    256.0          127         NaN   
256    257.0          127         NaN   
257    258.0          127         NaN   
258    259.0          127         NaN   
259    260.0          127         NaN   
260    261.0          127         NaN   
261    262.0          132         NaN   
262    263.0          132         NaN   
263    264.0          144         NaN   
264    265.0          144         NaN   
265    266.0          144         NaN   
266    267.0          144         NaN   
267    268.0          144         NaN   
268    269.0          110         NaN   
269    270.0          110         NaN   
270    271.0          110         NaN   
271    272.0          110         NaN   
272    273.0          110         NaN   
273    274.0          110         NaN   
274    275.0          110         NaN   
275    276.0          110         NaN   
276    277.0          113         NaN   
277    278.0          113         NaN   
278    279.0          113         NaN   
279    280.0          113         NaN   
280    281.0          113         NaN   
281    282.0          113         NaN   
282    283.0          116         NaN   
...      ...          ...         ...   
1863  1864.0          142         NaN   
1864  1865.0          142         NaN   
1865  1866.0          126         NaN   
1866  1867.0          103         NaN   
1867  1868.0          103         NaN   
1868  1869.0          103         NaN   
1869  1870.0          105         NaN   
1870  1871.0          105         NaN   
1871  1872.0          108         NaN   
1872  1873.0          108         NaN   
1873  1874.0          104         NaN   
1874  1875.0          104         NaN   
1875  1876.0          140         NaN   
1876  1877.0          121         NaN   
1877  1878.0          121         NaN   
1878  1879.0          101         NaN   
1879  1880.0          128         NaN   
1880  1881.0          102         NaN   
1881  1882.0          128         NaN   
1882  1883.0          141         NaN   
1883  1884.0          102         NaN   
1884  1885.0          135         NaN   
1885  1886.0          104         NaN   
1886  1887.0          133         NaN   
1887  1888.0          108         NaN   
1888  1889.0          126         NaN   
1889  1890.0          115         NaN   
1890  1891.0          108         NaN   
1891  1892.0          115         NaN   
1923  1924.0          246         NaN   

                                         Naziv ustanove          Tip ustanove  \
253            В.техничка школа струковних студија, Ниш  Државна висока школа   
254            В.техничка школа струковних студија, Ниш  Државна висока школа   
255            В.техничка школа струковних студија, Ниш  Државна висока школа   
256            В.техничка школа струковних студија, Ниш  Државна висока школа   
257            В.техничка школа струковних студија, Ниш  Државна висока школа   
258            В.техничка школа струковних студија, Ниш  Државна висока школа   
259            В.техничка школа струковних студија, Ниш  Државна висока школа   
260            В.техничка школа струковних студија, Ниш  Државна висока школа   
261   Висока школа струковних студија за васпитаче А...  Државна висока школа   
262   Висока школа струковних студија за васпитаче А...  Државна висока школа   
263   Висока техничка школа струковних студија - Зре...  Државна висока школа   
264   Висока техничка школа струковних студија - Зре...  Државна висока школа   
265   Висока техничка школа струковних студија - Зре...  Државна висока школа   
266   Висока техничка школа струковних студија - Зре...  Државна висока школа   
267   Висока техничка школа струковних студија - Зре...  Државна висока школа   
268   Висока здравствена школа струковних студија Бе...  Државна висока школа   
269   Висока здравствена школа струковних студија Бе...  Државна висока школа   
270   Висо

Да бисмо погледати број студената по различитим нивоима студирања, то можемо једноставно урадити користећи функцију **groupby** и сумирати вредносту колоне *Broj studenata* по групама:

In [18]:
studenti.groupby(studenti.Nivo)['Broj studenata'].sum()

Nivo
Докторске академске     11628.0
Интегр. академске       19623.0
Мастер академске        24006.0
Основне академске      140053.0
Основне струковне       42867.0
Спец. академске           718.0
Спец. струковне          3072.0
Name: Broj studenata, dtype: float64

За даљу анализу, фокусираћемо се само на основне и интегрисане студије (у којима су основне и мастер студије интегрисане):

In [19]:
studenti = studenti[(studenti.Nivo=='Интегр. академске')|(studenti.Nivo=='Основне академске')]
studenti = studenti.reset_index()

Број студената на последњој години студија:

In [20]:
stud_posl=[]
for index, row in studenti.iterrows():
    if ((index%10)==50):
        print(index)
    posl=-1
    if row['Dužina trajanja programa (u god)']==3:
        posl=row['Studenti treća godina']
    elif row['Dužina trajanja programa (u god)']==4:
        posl=row['Studenti četvrta godina']
    elif row['Dužina trajanja programa (u god)']==5:
        posl=row['Studenti peta godina']
    elif row['Dužina trajanja programa (u god)']==6:
        posl=row['Studenti šesta godina']
    else:
        print(row['Dužina trajanja programa (u god)'])
    
    if posl==-1:
        print(row['Dužina trajanja programa (u god)'],row['Naziv ustanove'])
    else:
        stud_posl.append(posl)
studenti['Broj studenata na zavrsnoj godini']=stud_posl

Ову табелу *studenti* која садржи све смерове основних и интегрисаних студија можемо сачувати у посебан .csv фајл уколико желимо детаљно све смерове да анализирамо:

In [21]:
studenti.to_csv('studenti_osnovnih_studija.csv',index=False)

Међутим, направићемо још једну верзију табеле, у којој ћемо агрегирати податке по установама и дужинама трајања програма (тј. ако неки факултет има више смерова који једнако дуго трају, воде истом звању и припадају истом пољу, сабираћемо укупан број студената на тим програмима). На тај начин заобилазимо проблем настао чињеницом да неки програрми настају и нестају, или само мењају имена. То ћемо опет урадити груписањем, овај пут користећи више колона:

In [22]:
studenti_agregirano = pd.DataFrame(studenti.groupby(['Naziv ustanove','Univerzitet','Nivo','Zvanje','Dužina trajanja programa (u god)','Polje_skraceno','Vlasnistvo'])[['Akreditaciona kvota','Broj studenata','Broj prijavljenih kandidata','Broj svršenih studenata tokom prethodne školske godine','Studenti prva godina','Studenti druga godina','Studenti treća godina','Studenti četvrta godina','Studenti peta godina','Studenti šesta godina','Broj budžetskih studenata', 'Broj samofinansirajućih studenata',
       'Broj stranih studenata','Broj studenata na zavrsnoj godini']].sum())
studenti_agregirano.head(2)

Akreditaciona kvota  \
Naziv ustanove             Univerzitet              Nivo              Zvanje                           Dužina trajanja programa (u god) Polje_skraceno Vlasnistvo                              
Агрономски факултет, Чачак Универзитет у Крагујевцу Основне академске Дипломирани инжењер пољопривреде 4                                TT             Државни факултет                 60.0   
                                                                      Дипломирани инжењер технологије  4                                TT             Државни факултет                 40.0   

                                                                                                                                                                         Broj studenata  \
Naziv ustanove             Univerzitet              Nivo              Zvanje                           Dužina trajanja programa (u god) Polje_skraceno Vlasnistvo                         
Агрономски факултет, Чачак Универзитет у Крагујевцу Основне академске Дипломирани инжењер пољопривреде 4                                TT             Државни факултет           174.0   
                                                                      Дипломирани инжењер технологије  4                                TT             Државни факултет            69.0   

                                                                                                                                                                         Broj prijavljenih kandidata  \
Naziv ustanove             Univerzitet              Nivo              Zvanje                           Dužina trajanja programa (u god) Polje_skraceno Vlasnistvo                                      
Агрономски факултет, Чачак Универзитет у Крагујевцу Основне академске Дипломирани инжењер пољопривреде 4                                TT             Државни факултет                         36.0   
                                                                      Дипломирани инжењер технологије  4                                TT             Државни факултет                         20.0   

                                                                                                                                                                         Broj svršenih studenata tokom prethodne školske godine  \
Naziv ustanove             Univerzitet              Nivo              Zvanje                           Dužina trajanja programa (u god) Polje_skraceno Vlasnistvo                                                                 
Агрономски факултет, Чачак Универзитет у Крагујевцу Основне академске Дипломирани инжењер пољопривреде 4                                TT             Државни факултет                                                 15        
                                                                      Дипломирани инжењер технологије  4                                TT             Државни факултет                                                  3        

                                                                                                                                                                         Studenti prva godina  \
Naziv ustanove             Univerzitet              Nivo              Zvanje                           Dužina trajanja programa (u god) Polje_skraceno Vlasnistvo                               
Агрономски факултет, Чачак Универзитет у Крагујевцу Основне академске Дипломирани инжењер пољопривреде 4                                TT             Државни факултет                  50.0   
                                                                      Дипломирани инжењер технологије  4                                TT             Државни факултет                  22.0   

                                                                                                                                                                        

Након инструкције груписања увек настаје нова табела која је индексирана по колонама које су коришћене за груписање (што видимо у претходном испису). Међутим, често ће нам бити једноставније да имамо једну индексну колону која садржи редне бројеве па ћемо искористити функцију **reset_index**:

In [23]:
studenti_agregirano = studenti_agregirano.reset_index()
studenti_agregirano.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 369 entries, 0 to 368
Data columns (total 21 columns):
Naziv ustanove                                            369 non-null object
Univerzitet                                               369 non-null object
Nivo                                                      369 non-null object
Zvanje                                                    369 non-null object
Dužina trajanja programa (u god)                          369 non-null int64
Polje_skraceno                                            369 non-null object
Vlasnistvo                                                369 non-null object
Akreditaciona kvota                                       369 non-null float64
Broj studenata                                            369 non-null float64
Broj prijavljenih kandidata                               369 non-null float64
Broj svršenih studenata tokom prethodne školske godine    369 non-null int64
Studenti prva godina                

Ова агрегирана табела има неке колоне у којима су текстуални подаци па ћемо и све вреднсоти у тим колонама променити у латинично писмо користећи код у наставку:

In [24]:
lista_kolona_za_izmenu = ['Naziv ustanove','Univerzitet','Nivo','Zvanje','Vlasnistvo']
for kolona in lista_kolona_za_izmenu:
    studenti_agregirano[kolona] = [cyrtranslit.to_latin(text,'sr') for text in list(studenti_agregirano[kolona])]
studenti_agregirano.head(2)

Naziv ustanove               Univerzitet               Nivo  \
0  Agronomski fakultet, Čačak  Univerzitet u Kragujevcu  Osnovne akademske   
1  Agronomski fakultet, Čačak  Univerzitet u Kragujevcu  Osnovne akademske   

                               Zvanje  Dužina trajanja programa (u god)  \
0  Diplomirani inženjer poljoprivrede                                 4   
1    Diplomirani inženjer tehnologije                                 4   

  Polje_skraceno        Vlasnistvo  Akreditaciona kvota  Broj studenata  \
0             TT  Državni fakultet                 60.0           174.0   
1             TT  Državni fakultet                 40.0            69.0   

   Broj prijavljenih kandidata  ...  Studenti prva godina  \
0                         36.0  ...                  50.0   
1                         20.0  ...                  22.0   

   Studenti druga godina  Studenti treća godina  Studenti četvrta godina  \
0                   29.0                   36.0                     28.0   
1                   19.0                   14.0                     10.0   

   Studenti peta godina  Studenti šesta godina  Broj budžetskih studenata  \
0                  31.0                    0.0                      132.0   
1                   4.0                    0.0                       39.0   

   Broj samofinansirajućih studenata  Broj stranih studenata  \
0                               42.0                     0.0   
1                               30.0                     0.0   

   Broj studenata na zavrsnoj godini  
0                               28.0  
1                               10.0  

[2 rows x 21 columns]

Када баратамо са подацима који настају ручним уношењем на више места (у овом случају на више различитих установа па се ти подаци даље обједињују), честе су грешке или неусаглашености. Загледањем табеле приметили смо један такав пример са Математичким факултетом који ћемо илустровати у наставку. Излистајмо за почетак све редове за које је назив установе Математички факултет:

In [25]:
studenti_agregirano[studenti_agregirano['Naziv ustanove']=='Matematički fakultet']

Naziv ustanove             Univerzitet               Nivo  \
63  Matematički fakultet  Univerzitet u Beogradu  Osnovne akademske   
64  Matematički fakultet  Univerzitet u Beogradu  Osnovne akademske   
65  Matematički fakultet  Univerzitet u Beogradu  Osnovne akademske   

                     Zvanje  Dužina trajanja programa (u god) Polje_skraceno  \
63     Diplomirani astronom                                 4             PM   
64  Diplomirani matematičar                                 4             PM   
65             Informatičar                                 4             PM   

          Vlasnistvo  Akreditaciona kvota  Broj studenata  \
63  Državni fakultet                 25.0            70.0   
64  Državni fakultet                250.0          1170.0   
65  Državni fakultet                160.0           717.0   

    Broj prijavljenih kandidata  ...  Studenti prva godina  \
63                        915.0  ...                  40.0   
64                        915.0  ...                 433.0   
65                        915.0  ...                 215.0   

    Studenti druga godina  Studenti treća godina  Studenti četvrta godina  \
63                   12.0                    6.0                     12.0   
64                  213.0                  178.0                    346.0   
65                  167.0                  182.0                    153.0   

    Studenti peta godina  Studenti šesta godina  Broj budžetskih studenata  \
63                   0.0                    0.0                       32.0   
64                   0.0                    0.0                      461.0   
65                   0.0                    0.0                      342.0   

    Broj samofinansirajućih studenata  Broj stranih studenata  \
63                               38.0                     0.0   
64                              708.0                     1.0   
65                              375.0                     0.0   

    Broj studenata na zavrsnoj godini  
63                               12.0  
64                              346.0  
65                              153.0  

[3 rows x 21 columns]

Видимо да има три различита усмерења смера основних студија - дипломирани астроном, математичар и информатичар. Видимо да су подаци по колонама углавном различити за ове смерове (што је и очекивано), али видимо да је број пријављених кандидата за сва три усмерења 915. То нам се чини да је маловероватно (то би значило да скоро 3000 студената годишње жели да упише Математички факултет), већ је вероватније да број пријављених кандидата није раздвојен по смеровима, већ је ово укупан број људи који су се пријавили за Математички факултет. Можемо запамтити ову информацију и узети је у обзир у тумачењу резултата у наставку, или решити да и остале податке по колонама агрегирамо и разматрамо заједно цео Математички факултет. Ту опцију ћемо приказати у наставку.

У речнику tekstualne_kolone сачуваћемо све податке који се налазе у текстуалним колонама у вези за Мат. факултетом:

In [26]:
tekstualne_kolone ={'Naziv ustanove':'Matematički fakultet','Univerzitet':'Univerzitet u Beogradu','Nivo':'Osnovne akademske','Zvanje':'Astronom/Matematičar/Informatičar','Dužina trajanja programa (u god)':4,'Polje_skraceno':'PM','Vlasnistvo':'Državni fakultet'}

Док ћемо у речнику numericke_kolone сачувати нумеричке податке до којих ћемо доћи агрегирањем, тј. сабирањем свих вредности по колонама за сва 3 усмерења на Мат. факултету:

In [27]:
numericke_kolone = dict(studenti_agregirano[studenti_agregirano['Naziv ustanove']=='Matematički fakultet'][['Akreditaciona kvota', 'Broj studenata', 'Broj prijavljenih kandidata',
       'Broj svršenih studenata tokom prethodne školske godine',
       'Studenti prva godina', 'Studenti druga godina',
       'Studenti treća godina', 'Studenti četvrta godina',
       'Studenti peta godina', 'Studenti šesta godina',
       'Broj budžetskih studenata', 'Broj samofinansirajućih studenata',
       'Broj stranih studenata', 'Broj studenata na zavrsnoj godini']].sum())

Објединичемо све ове податке у један речник:

In [28]:
tekstualne_kolone.update(numericke_kolone)
tekstualne_kolone

{'Naziv ustanove': 'Matematički fakultet',
 'Univerzitet': 'Univerzitet u Beogradu',
 'Nivo': 'Osnovne akademske',
 'Zvanje': 'Astronom/Matematičar/Informatičar',
 'Dužina trajanja programa (u god)': 4,
 'Polje_skraceno': 'PM',
 'Vlasnistvo': 'Državni fakultet',
 'Akreditaciona kvota': 435.0,
 'Broj studenata': 1957.0,
 'Broj prijavljenih kandidata': 2745.0,
 'Broj svršenih studenata tokom prethodne školske godine': 125.0,
 'Studenti prva godina': 688.0,
 'Studenti druga godina': 392.0,
 'Studenti treća godina': 366.0,
 'Studenti četvrta godina': 511.0,
 'Studenti peta godina': 0.0,
 'Studenti šesta godina': 0.0,
 'Broj budžetskih studenata': 835.0,
 'Broj samofinansirajućih studenata': 1121.0,
 'Broj stranih studenata': 1.0,
 'Broj studenata na zavrsnoj godini': 511.0}

На овај начин смо агрегирали и податке у колони од које је све кренуло - број пријављених студената - па треба променити ту вредност:

In [29]:
tekstualne_kolone['Broj prijavljenih kandidata']=tekstualne_kolone['Broj prijavljenih kandidata']/3
tekstualne_kolone

{'Naziv ustanove': 'Matematički fakultet',
 'Univerzitet': 'Univerzitet u Beogradu',
 'Nivo': 'Osnovne akademske',
 'Zvanje': 'Astronom/Matematičar/Informatičar',
 'Dužina trajanja programa (u god)': 4,
 'Polje_skraceno': 'PM',
 'Vlasnistvo': 'Državni fakultet',
 'Akreditaciona kvota': 435.0,
 'Broj studenata': 1957.0,
 'Broj prijavljenih kandidata': 915.0,
 'Broj svršenih studenata tokom prethodne školske godine': 125.0,
 'Studenti prva godina': 688.0,
 'Studenti druga godina': 392.0,
 'Studenti treća godina': 366.0,
 'Studenti četvrta godina': 511.0,
 'Studenti peta godina': 0.0,
 'Studenti šesta godina': 0.0,
 'Broj budžetskih studenata': 835.0,
 'Broj samofinansirajućih studenata': 1121.0,
 'Broj stranih studenata': 1.0,
 'Broj studenata na zavrsnoj godini': 511.0}

Коначно, можемо овај речник додати као један нови ред у табелу коју посматрамо:

In [30]:
studenti_agregirano = studenti_agregirano.append(tekstualne_kolone,ignore_index=True)

In [31]:
studenti_agregirano[studenti_agregirano['Naziv ustanove']=='Matematički fakultet']

Naziv ustanove             Univerzitet               Nivo  \
63   Matematički fakultet  Univerzitet u Beogradu  Osnovne akademske   
64   Matematički fakultet  Univerzitet u Beogradu  Osnovne akademske   
65   Matematički fakultet  Univerzitet u Beogradu  Osnovne akademske   
369  Matematički fakultet  Univerzitet u Beogradu  Osnovne akademske   

                                Zvanje  Dužina trajanja programa (u god)  \
63                Diplomirani astronom                                 4   
64             Diplomirani matematičar                                 4   
65                        Informatičar                                 4   
369  Astronom/Matematičar/Informatičar                                 4   

    Polje_skraceno        Vlasnistvo  Akreditaciona kvota  Broj studenata  \
63              PM  Državni fakultet                 25.0            70.0   
64              PM  Državni fakultet                250.0          1170.0   
65              PM  Državni fakultet                160.0           717.0   
369             PM  Državni fakultet                435.0          1957.0   

     Broj prijavljenih kandidata  ...  Studenti prva godina  \
63                         915.0  ...                  40.0   
64                         915.0  ...                 433.0   
65                         915.0  ...                 215.0   
369                        915.0  ...                 688.0   

     Studenti druga godina  Studenti treća godina  Studenti četvrta godina  \
63                    12.0                    6.0                     12.0   
64                   213.0                  178.0                    346.0   
65                   167.0                  182.0                    153.0   
369                  392.0                  366.0                    511.0   

     Studenti peta godina  Studenti šesta godina  Broj budžetskih studenata  \
63                    0.0                    0.0                       32.0   
64                    0.0                    0.0                      461.0   
65                    0.0                    0.0                      342.0   
369                   0.0                    0.0                      835.0   

     Broj samofinansirajućih studenata  Broj stranih studenata  \
63                                38.0                     0.0   
64                               708.0                     1.0   
65                               375.0                     0.0   
369                             1121.0                     1.0   

     Broj studenata na zavrsnoj godini  
63                                12.0  
64                               346.0  
65                               153.0  
369                              511.0  

[4 rows x 21 columns]

Сада имамо четири реда која одговарају Математичком факултету, па ћемо у наставку избацити редове који су претходно постојали и фокусирати се само на агрегирани:

In [32]:
studenti_agregirano = studenti_agregirano.drop([63,64,65])
studenti_agregirano[studenti_agregirano['Naziv ustanove']=='Matematički fakultet']

Naziv ustanove             Univerzitet               Nivo  \
369  Matematički fakultet  Univerzitet u Beogradu  Osnovne akademske   

                                Zvanje  Dužina trajanja programa (u god)  \
369  Astronom/Matematičar/Informatičar                                 4   

    Polje_skraceno        Vlasnistvo  Akreditaciona kvota  Broj studenata  \
369             PM  Državni fakultet                435.0          1957.0   

     Broj prijavljenih kandidata  ...  Studenti prva godina  \
369                        915.0  ...                 688.0   

     Studenti druga godina  Studenti treća godina  Studenti četvrta godina  \
369                  392.0                  366.0                    511.0   

     Studenti peta godina  Studenti šesta godina  Broj budžetskih studenata  \
369                   0.0                    0.0                      835.0   

     Broj samofinansirajućih studenata  Broj stranih studenata  \
369                             1121.0                     1.0   

     Broj studenata na zavrsnoj godini  
369                              511.0  

[1 rows x 21 columns]

Коначно, сачуваћемо и овај фајл за даљу анализу и визуализацију:

In [33]:
studenti_agregirano.to_csv('studenti_osnovnih_studija_agregirano.csv',index=False)